Modify site switch to use site str()

JackMorganNZ opened the pull request:
## Description

The site switcher uses a custom string representation, which duplicates most of the logic from `Site.__str__()`. This switcher is shown on site settings pages.

This change reduces the code repetition, and improving the readability of choices for users. Other packages, such as wagtail menus, show both the site name and domain in their site switcher.

I've updated the tests to check the string representation is used, and that all sites are listed as options. The existing test checking order of sites has been removed, as I cannot find any reasoning for it, I can restore it if required.

### Useful links

- [Wagtail Site model .__str__() method](https://github.com/wagtail/wagtail/blob/23408a513f4b6754cea6bd0563dae4577dc68e3f/wagtail/models/sites.py#L122-L131)

### Screenshots

Before:

<img width="1696" height="334" alt="image" src="https://github.com/user-attachments/assets/5db27102-93a3-49aa-8b4c-cb327341e52f" />

After:

<img width="1696" height="336" alt="image" src="https://github.com/user-attachments/assets/89a274c4-6e2e-4433-8d8a-8bd5b67de913" />

JackMorganNZ committed:
Order sites by site_name falling back to hostname

lb- commented:
Thanks for the PR @JackMorganNZ - I've flagged for a review and set the full CI to run. 

Could you update the description with a link to the code where the other implementation of the model's string representation is. I know we can find it but it just helps for reviewing.

JackMorganNZ commented:
Thanks @lb-, I've added the link to the original method (should have thought of that the first time!).

gasman reviewed (changes requested):
Thanks @JackMorganNZ!

It looks like the original logic for this listing came from #6004 (which in turn was a follow-up to #5891), and the rationale was something like:

* `site_name` is an optional field that users may not have filled in, so out of `site_name` and `hostname`, it's better to order by hostname
* We don't want to order by a field we're not showing, so that implies displaying hostname rather than the standard `str` representation

With that in mind, I don't think we should go ahead with this change unless we also account for the ordering. I think it should be possible to arrive at an ordering on site name falling back on hostname with some careful use of database functions - [`Coalesce`](https://docs.djangoproject.com/en/5.2/ref/models/database-functions/#coalesce) is _almost_ what we need, but that only works if we're storing NULL for an empty site_name, and I'm pretty sure we use an empty string instead. A [conditional expression](https://docs.djangoproject.com/en/5.2/ref/models/conditional-expressions/) is a possibility, but I dare say [`Concat`](https://docs.djangoproject.com/en/5.2/ref/models/database-functions/#concat)`('site_name', 'hostname')` is close enough to do the job.

laymonage commented:
Hi @JackMorganNZ, are you still interested in pursuing this? Thanks!

JackMorganNZ commented:
> Hi @JackMorganNZ, are you still interested in pursuing this? Thanks!

@laymonage Sorry, totally slipped my list. Will do it this weekend.

JackMorganNZ commented:
I've updated the PR with the requested ordering change (option 1). Rather than `Concat('site_name', 'hostname')`, I went with a conditional expression as you also suggested, since it helps in an edge case:

```python
.order_by(
    Case(
        When(site_name="", then=Lower("hostname")),
        default=Lower("site_name"),
    ),
    Lower("hostname"),
)
```

The reason Concat falls short is where a site named "be" (hostname ar.com) and a nameless bear.com would produce identical sort keys under Concat ("bear.com" for both), making their relative order non-deterministic. With Case/When the keys are "be" and "bear.com", therefore distinct and deterministic. The secondary Lower("hostname") acts as a tiebreaker for sites sharing a site_name.

Updated tests to reflect the new ordering and added coverage for different ordering permutations.

The backend tests on CircleCI are failing for what appears to be an unrelated reason.

laymonage commented on wagtail/contrib/settings/tests/site_specific/test_forms.py:
```diff
@@ -5,27 +5,53 @@
+            label_by_id[site_id] = label
+
+        for site in (site_1, site_2, site_3, site_4):
+            self.assertEqual(label_by_id[site.id], str(site))
```
I agree with Matt's comment here. I'll make that change as I merge.

laymonage reviewed (commented):
Changes look good to me! Just a minor tweak to the tests that I can do as I merge. Thanks!

JackMorganNZ commented:
Thanks @laymonage, totally missed that comment when reviewing (definitely not overtired from an infant... 😐 ). Thanks for getting this over the line! 🎉
